### DISCUSSION TITLE
Possible issue with selects when using relationships

### DISCUSSION THREAD

**Discussion by liamoreilly** [spatie/laravel-html#211] [General]
Hi,

Let's say you are building an edit user page and want a multiselect for the roles a user has using the spatie/laravel-permission package (but it need not be this package, this issue would be the same with any relationship).

You provide the options array to the `html->multiselect` as an associative array of the form role ID to role name using some controller code of the form:
```
$options = Role::orderBy('name')->get()->pluck('name', 'id')->toArray();
```
Now let's say you have a `roles` relationship on your user model. In your create view you open a model form with this user as the model.
```
{{ html()->modelForm($user, 'PATCH', route('users.update', [$user->id]))->open() }}
{{ html()->multiselect('roles', $options) }}
{{ html()->form()->close() }}
```

One would expect the multiselect to be prepopulated with the correct roles for the user. However, this fails currently.

The issue boils down to [Line 129 of laravel-html/tree/main/src/Elements/Select.php](https://github.com/spatie/laravel-html/blob/07446a1a944335bbcb0b9d68d67f443fdd97a5e5/src/Elements/Select.php#L129C1-L129C49). Here, `$this->value` is an Eloquent Collection (`Illuminate\Database\Eloquent\Collection`), emphasis on the Eloquent aspect. This collection of role models has been correctly pulled in from the user model. After this line executes, `$value` is a normal Collection (`Illuminate\Support\Collection`). Later on in [Line 146](https://github.com/spatie/laravel-html/blob/07446a1a944335bbcb0b9d68d67f443fdd97a5e5/src/Elements/Select.php#L146) the contains method is called on this normal collection. This method behaves differently if the collection is a normal one or an Eloquent one. Each option's value is an integer/string. This integer/string fails to match the models in normal collections but it would match in Eloquent collections.

As a note, 'old' values from failed validation work just fine as the `$value` collection is then a list of integer IDs. Only when retrieving them from the model are they a Collection of Role models.

If [Line 129](https://github.com/spatie/laravel-html/blob/07446a1a944335bbcb0b9d68d67f443fdd97a5e5/src/Elements/Select.php#L129C1-L129C49) was wrapped in an "if not instance of Collection" then I suspect there would be no breakages, but this scenario would work out - there would be no need to turn it into a Collection if it is already a Collection and an Eloquent Collection would remain as one.

As a work around I could imagine creating a new attribute named something like `role_ids` and working with these, but this feels a bit dirty - for each occurrence of this scenario a separate attribute has to be created in the models.

I recall this matching working as an Eloquent collection in laravelcollective/html. Either way, it makes this use-case particularly problematic (i.e., using a relationship with select/multiselect).

Any ideas? Maybe I am doing something fundamentally wrong here, but I currently don't see it.

Thanks,
Liam

**Comment by SwanseaBen** [comment]
Would like to know to!

**Comment by JuanEduardoGT** [comment]
I have the same problem.

**Comment by nikosv** [comment]
i just sent PR https://github.com/spatie/laravel-html/pull/237 with the solution proposed by @liamoreilly

**Reply by SwanseaBen** [reply]
Have not tested it but thanks
